### PR TITLE
hostapp: allow devices to provision & set target state via POST

### PR DESCRIPTION
**Pull request by xginn8**
Change-type: patch
Signed-off-by: Matthew McGinn <matthew@balena.io>

### CODE REVIEW

File: src/features/hostapp/hooks/target-hostapp.ts [hunk 1]
```diff
@@ -55,7 +58,46 @@ hooks.addPureHook('PATCH', 'resin', 'device', {
...
+			args.request.values.os_variant != null
+		) {
+			let deviceType;
+			if (!args.request.values.is_of__device_type) {
+				[deviceType] = await args.api.get({
+					resource: 'device_type',
+					options: {
+						$select: 'id',
+						$filter: {
+							slug: { $eq: args.request.values.device_type },
```
Comment (Page-) [inline src/features/hostapp/hooks/target-hostapp.ts:76]: `$eq` is normally unnecessary:
```suggestion
							slug: args.request.values.device_type,
```

File: src/features/hostapp/hooks/target-hostapp.ts [hunk 2]
```diff
@@ -55,7 +58,46 @@ hooks.addPureHook('PATCH', 'resin', 'device', {
...
+
+			const hostappRelease = await getOSReleaseResource(
+				args.api,
+				args.request.values.os_version,
+				args.request.values.os_variant,
+				args.request.values.is_of__device_type ?? deviceType?.id,
+			);
+			if (hostappRelease && hostappRelease.length > 0) {
+				args.request.values.should_be_operated_by__release =
+					hostappRelease[0].id;
```
Comment (Page-) [inline src/features/hostapp/hooks/target-hostapp.ts:93]: Can the `PATCH` hook do this the same way?

Comment (xginn8) [reply src/features/hostapp/hooks/target-hostapp.ts:93]: the `PATCH` hook also enforces some extra constraints (not invalidated and not a downgrade), whereas the `POST` is less picky. do you mean specifically how the return value of `getOSReleaseResource` is used and validated?

Comment (Page-) [reply src/features/hostapp/hooks/target-hostapp.ts:93]: Yeah, the other bits can be specific to the PATCH as need be but this bit of setting the default seems like it should be the same for both?

File: src/features/hostapp/hooks/target-hostapp.ts [hunk 3]
```diff
@@ -55,7 +58,46 @@ hooks.addPureHook('PATCH', 'resin', 'device', {
...
+						$select: 'id',
+						$filter: {
+							slug: args.request.values.device_type,
+						},
+					},
+				});
+			}
+			if (!deviceType) {
+				return;
+			}
```
Comment (Page-) [inline src/features/hostapp/hooks/target-hostapp.ts:83]: Oh I just realized I made a mistake here, this'll break if there's no match
```suggestion
				const [ dt ] = await args.api.get({
					resource: 'device_type',
					options: {
						$select: 'id',
						$filter: {
							slug: args.request.values.device_type,
						},
					},
				});
			}
			if (!dt) {
				return;
			}
			deviceType = dt.id
```

**Comment by xginn8** [comment]
@balena-ci I self-certify!

**Pull request merged**